python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -25,6 +26,135 @@
+        self._worker.check_connected()
+
+        ctx = ChannelContext.get_current()
+        self.serialization_ctx = ctx.serialization_context
+        assert self.serialization_ctx is not None
+
+    @classmethod
+    def from_channels(
+        cls,
+        gpu_data_channel: "TorchTensorNcclChannel",
+        cpu_data_channel: Optional["Channel"],
+    ):
+        return cls(
+            writer=None,
+            readers=None,
+            gpu_data_typ=None,
+            cpu_data_typ=None,
+            _gpu_data_channel=gpu_data_channel,
+            _cpu_data_channel=cpu_data_channel,
+        )
+
+    def __reduce__(self):
+        return (
+            NestedTorchTensorNcclChannel.from_channels,
+            (self._gpu_data_channel, self._cpu_data_channel),
+        )
+
+    def ensure_registered_as_writer(self):
+        self._gpu_data_channel.ensure_registered_as_writer()
+        if self._cpu_data_channel is not None:
+            self._cpu_data_channel.ensure_registered_as_writer()
+
+    def ensure_registered_as_reader(self):
+        self._gpu_data_channel.ensure_registered_as_reader()
+        if self._cpu_data_channel is not None:
+            self._cpu_data_channel.ensure_registered_as_reader()
+
+    def write(self, value: Any):
+        self.serialization_ctx.reset_tensors([])
+        # All tensors found in `value` will be transferred via NCCL.
+        self.serialization_ctx.set_use_external_transport(True)
+
+        try:
+            serialized_cpu_data = self._worker.get_serialization_context().serialize(
+                value
+            )
+        except TypeError as e:
+            sio = io.StringIO()
+            ray.util.inspect_serializability(value, print_file=sio)
+            msg = (
+                "Could not serialize the put value "
+                f"{repr(value)}:\n"
+                f"{sio.getvalue()}"
+            )
+            raise TypeError(msg) from e
+        finally:
+            # Pop the tensors that were found during serialization of `value`.
+            tensors_to_send = self.serialization_ctx.reset_tensors([])
+            # Reset the serialization method to now serialize torch.Tensors
+            # normally.
+            self.serialization_ctx.set_use_external_transport(False)
+
+        self._gpu_data_channel.write(tensors_to_send)
+        self._cpu_data_channel.write(serialized_cpu_data)
+
+    def begin_read(self) -> Any:
+        tensors = self._gpu_data_channel.begin_read()
+
+        if self._gpu_data_channel.has_static_type():
+            # If the channel was declared with a static TorchTensorType, then
+            # the task is allowed to return at most one tensor, and its shape
+            # and dtype must match the declared type. Wrap the tensor in a
+            # list since the following calls expect a list.
+            tensors = [tensors]
+
+        self.serialization_ctx.reset_tensors(tensors)
+        data = self._cpu_data_channel.begin_read()
```

Comment:
this part of the flow is a little confusing. Can you comment why we are doing this? I assume this self.serialization_ctx is used internally at cpu_data_channel?

Reply:
The serialization protocol for `class NestedTorchTensorNcclChannel` is as follows:

* The sender worker process will call `NestedTorchTensorNcclChannel.write()`, and the receiver worker process will call `NestedTorchTensorNcclChannel.begin_read()`.

* For example, we have a object `my_obj` as follows:
  ```python
  class MyObj:
    self.tensor = t # Torch.Tensor(...)
    self.msg = "123"
  ```

* The sender process calls `NestedTorchTensorNcclChannel.write(my_obj)`.
  ```python
  serialized_cpu_data = self._worker.get_serialization_context().serialize(
      value
  )
  ...
  tensors_to_send = self.serialization_ctx.reset_tensors([])
  ...
  self._gpu_data_channel.write(tensors_to_send)
  self._cpu_data_channel.write(serialized_cpu_data)
  ``` 
  * The `serialize` function will call `serialize_tensor` for all tensors in value (i.e., my_obj). The function `serialize_tensor` will do two things: (1) store the tensor in the serialization context, and (2) use the index (an `int`) in `serialization_ctx.tensors` as a placeholder in `serialized_cpu_data`.
  * `serialized_cpu_data` will look like:
    ```python
    class MyObj:
       self.tensor = serialize_tensor(t) # => an index => serialization_ctx.tensors[serialize_tensor(t)] is t
       self.msg = serialize_str("123")
    ```
  * `tensors_to_send` is `serialization_ctx.tensors`.

* The receiver process calls `NestedTorchTensorNcclChannel.begin_read()`.
   ```python
   tensors = self._gpu_data_channel.begin_read()
   ...
   self.serialization_ctx.reset_tensors(tensors)
   data = self._cpu_data_channel.begin_read()
   ```
   * `tensors` are the tensors sent by `self._gpu_data_channel.write(tensors_to_send)` which is sender's `serialization_ctx.tensors`.
   * Use `self.serialization_ctx.reset_tensors(tensors)` to set receiver's `serialization_ctx.tensors` to `tensors`.
   * `data = self._cpu_data_channel.begin_read()`
     ```python
     class MyObj:
       # i.e. serialization_ctx.tensors[serialize_tensor(t)]
       self.tensor = deserialize_tensor(serialize_tensor(t)) 
       self.msg = deserialize_str(serialize_str("123"))
     ```
     * Because `self._cpu_data_channel.begin_read()` depends on `serialization_ctx.tensors`, we must call `self.serialization_ctx.reset_tensors(tensors)` before `begind_read()`.